--- a/crypto/api.c
+++ b/crypto/api.c
@@ -204,14 +204,17 @@ static struct crypto_alg *crypto_larval_
 	struct crypto_larval *larval;
 	long time_left;
 
+	pr_info("crypto_larval_wait: Waiting for %s...", alg->cra_name);
 again:
 	larval = container_of(alg, struct crypto_larval, alg);
 
+	pr_info("crypto_larval_wait: larval waiter (%s): %p", alg->cra_name, &larval->completion);
+
 	if (!crypto_boot_test_finished())
 		crypto_start_test(larval);
 
 	time_left = wait_for_completion_killable_timeout(
-		&larval->completion, 60 * HZ);
+		&larval->completion, 120 * HZ);
 
 	alg = larval->adult;
 	if (time_left < 0)
@@ -247,6 +250,12 @@ again:
 	if (!IS_ERR(alg) && crypto_is_larval(alg))
 		goto again;
 
+	if (IS_ERR(alg)) {
+		pr_info("crypto_larval_wait: returning error %d...", *(int *)alg);
+	} else {
+		pr_info("crypto_larval_wait: returning %s...", alg ? alg->cra_name : "none");
+	}
+	
 	return alg;
 }
 
@@ -356,7 +365,6 @@ struct crypto_alg *crypto_alg_mod_lookup
 		return larval;
 
 	ok = crypto_probing_notify(CRYPTO_MSG_ALG_REQUEST, larval);
-
 	if (ok == NOTIFY_STOP)
 		alg = crypto_larval_wait(larval, type, mask);
 	else {
@@ -712,6 +720,7 @@ void crypto_req_done(void *data, int err
 		return;
 
 	wait->err = err;
+	pr_info("crypto_req_done: %p", &wait->completion);
 	complete(&wait->completion);
 }
 EXPORT_SYMBOL_GPL(crypto_req_done);
--- a/crypto/algapi.c
+++ b/crypto/algapi.c
@@ -239,6 +239,8 @@ static void crypto_alg_finish_registrati
 {
 	struct crypto_alg *q;
 
+	pr_info("crypto_alg_finish_registration: %s(%s) - %u", alg->cra_name, alg->cra_driver_name, alg->cra_flags);
+	
 	list_for_each_entry(q, &crypto_alg_list, cra_list) {
 		if (q == alg)
 			continue;
@@ -302,6 +304,8 @@ __crypto_register_alg(struct crypto_alg
 
 	INIT_LIST_HEAD(&alg->cra_users);
 
+	pr_info("__crypto_register_alg: %s(%s) - %u", alg->cra_name, alg->cra_driver_name, alg->cra_flags);
+
 	ret = -EEXIST;
 
 	list_for_each_entry(q, &crypto_alg_list, cra_list) {
